### PR TITLE
fix: align kube-rs with client-go config parsing

### LINKED ISSUE #1076: Parsing of kubeconfig is more strict than client-go
### Current and expected behavior

I noticed that `Kubeconfig::read()` is a bit more strict than client-go's when parsing the kubeconfig. For example:

In client-go, it's possible to have a context without a cluster or user. It won't actually work if you try to use that context, but it doesn't fail to parse it, meaning that users can use the other contexts that have all the required fields.

These are the differences I found:
- the `contexts.context` is optional in client-go, mandatory in kube-rs.
- the `contexts.context.name` is optional in client-go, mandatory in kube-rs.
- the `contexts.context.user` is optional in client-go, mandatory in kube-rs.
- the `users.user.auth-provider.config` is optional in client-go, mandatory in kube-rs.
- the `users.user.auth-provider.config` can be set to empty, `~` or null in client-go, but not in kube-rs
- there might be more though

### Possible solution

Relax the parsing of kubeconfig to match client-go as much as possible. To avoid further breaking changes, ideally if those fields are missing on kubeconfig, kube-rs would default it to empty strings / empty hash maps.

### Additional context

_No response_

### Environment

Client Version: v1.25.3
Kustomize Version: v4.5.7
Server Version: v1.22.8
WARNING: version difference between client (1.25) and server (1.22) exceeds the supported minor version skew of +/-1

### Configuration and features

```toml
kube = { version = "0.76.0", features = ["client", "runtime", "ws", "gzip"]  }
k8s-openapi = { version = "0.16.0", features = ["v1_24"] }
```

### Affected crates

kube-client

### Would you like to work on fixing this bug?

yes

### PR DESCRIPTION
**Pull request by goenning**
Signed-off-by: goenning <me@goenning.net>

## Motivation

Fixes #1076

## Solution

I went through each field on my kubeconfig and removed it / set it to null, then I tried both kubectl and cube-rs example kubectl. If it succeeded on kubectl but failed on kube-rs, I added a default and moved to the next field.

The part I struggled with was parsing of null values, such as 

```
  - cluster:
    name: gke_oenning-sandbox_us-central1-c_cluster-1
```

I couldn't find how to do it with serve alone, so I had to add the serde_with crate.

Some properties like `cluster.server` are useless when empty, but the user might not be interacting with that cluster/context, so it should not be causing errors for something that might not impact its usage. If they try to use that cluster, they'll then have an error somewhere else. 

Note: I'll add a unit test as well, but I want to get it out here for feedback early on in case someone disagree with the solution.

### CODE REVIEW

File: kube-client/Cargo.toml [hunk 1]
```diff
@@ -42,6 +42,7 @@ dirs = { package = "dirs-next", optional = true, version = "2.0.0" }
 serde = { version = "1.0.130", features = ["derive"] }
 serde_json = "1.0.68"
 serde_yaml = { version = "0.8.21", optional = true }
+serde_with = { version = "2.0.1" }
```
Comment (clux) [inline kube-client/Cargo.toml:45]: i think this is fine to add.. it's a [fairly popular crate](https://crates.io/crates/serde_with), but you might be able to turn off a lot of [its features](https://github.com/jonasbb/serde_with/blob/master/serde_with/Cargo.toml#L26-L59)?

File: kube-client/src/config/file_config.rs [hunk 2]
```diff
@@ -72,20 +73,24 @@ pub struct NamedExtension {
...
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
 #[cfg_attr(test, derive(PartialEq, Eq))]
 pub struct NamedCluster {
     /// Name of cluster
+    #[serde(default)]
     pub name: String,
     /// Information about how to communicate with a kubernetes cluster
+    #[serde_as(as = "serde_with::DefaultOnNull")]
     pub cluster: Cluster,
```
[the remark this answers is not in the document]
Comment (Dav1dde) [reply kube-client/src/config/file_config.rs:85]: I think using an `Option` is a lot clearer and avoids a dependency 👍

[the remark this answers is not in the document]
Comment (goenning) [reply kube-client/src/config/file_config.rs:85]: I opted for the DefaultOnNull to avoid breaking changes. Also from a consumer perspective, I find working with non-null structs much simpler and overall better experience, especially these ones that are rarely null.

That's just my opinion, but I'm easy, if majority prefers to use Option and remove the dependency, I can change it 👍

[the remark this answers is not in the document]
Comment (clux) [reply kube-client/src/config/file_config.rs:85]: I am a bit torn on it myself. It's not a dependency that is in our recursive dependency tree so it will be a new dependency for many. If it's ultimately just a slight inconvenience to have to `.unwrap_or_default` during construction phase for us internally (and for the small amount of people doing manual work on our structs), then it is possibly worth keeping defaulting and parsing separate (i.e. using options + unwrap_or_default).

Anyone else have opinions? Maybe @kazk ?

[the remark this answers is not in the document]
Comment (kazk) [reply kube-client/src/config/file_config.rs:85]: I think we should use `Option<T>`.

- Defaulting on null/missing is a surprising behavior in Rust, at least to me. It's more idiomatic to use `Option<T>`.
- `Option<T>` is easier to check if the value is present when we need them. We can also return a better error when missing. If we default on missing, we can't tell if it was missing in the config or default was provided.
- Changing to `Option<T>` is a breaking change, but it's a small and straightforward change. It should also be mostly internal.

[the remark this answers is not in the document]
Comment (goenning) [reply kube-client/src/config/file_config.rs]: Thanks all for the feedback, I changed to Option<T> now and removed serde_with.

File: kube-client/src/config/file_config.rs [hunk 3]
```diff
@@ -66,26 +66,30 @@ pub struct Preferences {
...
-#[derive(Clone, Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
 #[cfg_attr(test, derive(PartialEq, Eq))]
 pub struct NamedCluster {
     /// Name of cluster
+    #[serde(default)]
     pub name: String,
     /// Information about how to communicate with a kubernetes cluster
-    pub cluster: Cluster,
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (Dav1dde) [inline kube-client/src/config/file_config.rs:83]: Shouldn't this be:

```
#[serde(default, skip_serializing_if = "Option::is_none")]
```

Comment (Dav1dde) [reply kube-client/src/config/file_config.rs:83]: Yeah you're right it doesn't make a difference, unfortunately I didnt find anything in the docs about this. But the expanded macro code handles options differently and implicitly works like if there was a `default` applied to it. 👍

File: kube-client/src/config/file_loader.rs [hunk 4]
```diff
@@ -91,8 +91,8 @@ impl ConfigLoader {
 
         Ok(ConfigLoader {
             current_context: current_context.clone(),
-            cluster: cluster.clone(),
-            user: user.clone(),
+            cluster: cluster.clone().unwrap_or_default(),
+            user: user.clone().unwrap_or_default(),
```
Comment (kazk) [inline kube-client/src/config/file_loader.rs:95]: I think we should do the following instead of `unwrap_or_default()`:

```rust
        let cluster = config
            .clusters
            .iter()
            .find(|named_cluster| &named_cluster.name == cluster_name)
            // change `map` to `and_then` and clone
            .and_then(|named_cluster| named_cluster.cluster.clone())
            .ok_or_else(|| KubeconfigError::LoadClusterOfContext(cluster_name.clone()))?;

        let user_name = user.unwrap_or(&current_context.user);
        let user = config
            .auth_infos
            .iter()
            .find(|named_user| &named_user.name == user_name)
            // change `map` to `and_then` and clone
            .and_then(|named_user| named_user.auth_info.clone())
            .ok_or_else(|| KubeconfigError::FindUser(user_name.clone()))?;

        Ok(ConfigLoader {
            current_context: current_context.clone(),
            // remove clone
            cluster,
            user,
        })
```

Comment (goenning) [reply kube-client/src/config/file_loader.rs]: Indeed, that's a lot cleaner, thanks your suggestion! I've pushed the change.

File: kube-client/src/config/file_config.rs [hunk 5]
```diff
@@ -66,26 +66,30 @@ pub struct Preferences {
...
+    pub cluster: Option<Cluster>,
 }
 
 /// Cluster stores information to connect Kubernetes cluster.
-#[derive(Clone, Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
 #[cfg_attr(test, derive(PartialEq, Eq))]
 pub struct Cluster {
     /// The address of the kubernetes cluster (https://hostname:port).
+    #[serde(default)]
```
Comment (kazk) [inline kube-client/src/config/file_config.rs:92]: I think these new optional fields should align with existing optional fields and use `Option` + skip serializing if none.

Comment (goenning) [reply kube-client/src/config/file_config.rs:92]: Yeah that's reasonable. Just to be sure I understood it correctly: all strings with `default` attribute would become `Option<String>`?

Comment (kazk) [reply kube-client/src/config/file_config.rs:92]: Yeah. Also, `#[serde(skip_serializing_if = "Option::is_none")]`, just like other existing optional fields.

Comment (goenning) [reply kube-client/src/config/file_config.rs]: Done. This PR got bigger than I initially expected, but it seems like everything still works. As I'm new to Rust, I may have not implemented it correctly or might have some inefficiencies, so I appreciate if you could review it again. 

But this config: 
```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: ...
    server: https://kubernetes.docker.internal:6443
  name:
contexts:
- context:
    cluster: 
    user: 
  name: 
current-context:
kind: Config
preferences: {}
users:
- name:
  user:
    client-certificate-data: ...
    client-key-data: ...
```

Now works on both kubectl and kube-rs. I may have gone too far with this so I'm option for feedback If that's something that should be reverted.

File: kube-client/src/config/file_config.rs [hunk 6]
```diff
@@ -688,8 +708,8 @@ users:
         let cfg = Kubeconfig::from_yaml(config_yaml)?;
 
         // Ensure we have data from both documents:
-        assert_eq!(cfg.clusters[0].name, "k3d-promstack");
-        assert_eq!(cfg.clusters[1].name, "k3d-k3s-default");
+        assert_eq!(cfg.clusters[0].name.as_ref().unwrap(), "k3d-promstack");
+        assert_eq!(cfg.clusters[1].name.as_ref().unwrap(), "k3d-k3s-default");
```
Comment (clux) [inline kube-client/src/config/file_config.rs:712]: this stuff expands into something pretty big after running `just fmt`, maybe factor out `let cluster1 = config.clusters[1].cluster.as_ref().unwrap();` (and run `just fmt` having nightly rust available to fix ci)

File: kube-client/src/config/file_config.rs (hunk 5 above)
Comment (clux) [reply kube-client/src/config/file_config.rs]: It's looking pretty clean to me. Thanks a lot. Only left a nit on fmt things which is failing atm.

File: kube-client/src/config/file_config.rs (hunk 6 above)
Comment (goenning) [reply kube-client/src/config/file_config.rs:712]: Fixed it, but fmt didn't touch that line, it was on 657, I guess that's the one you were referring to?

File: kube-client/src/config/file_config.rs [hunk 7]
```diff
@@ -218,8 +225,10 @@ impl PartialEq for AuthInfo {
 #[cfg_attr(test, derive(PartialEq, Eq))]
 pub struct AuthProviderConfig {
     /// Name of the auth provider
+    #[serde(default)]
```
Comment (kazk) [inline kube-client/src/config/file_config.rs:228]: These should be changed too.

File: kube-client/src/config/mod.rs [hunk 8]
```diff
@@ -299,6 +295,8 @@ impl Config {
         let cluster_url = loader
             .cluster
             .server
+            .clone()
+            .unwrap_or_default()
```
Comment (kazk) [inline kube-client/src/config/mod.rs:299]: This should return an error for missing cluster url.

File: kube-client/src/config/file_loader.rs [hunk 9]
```diff
@@ -59,40 +59,42 @@ impl ConfigLoader {
...
     ) -> Result<Self, KubeconfigError> {
+        let current_context_name = &config.current_context.clone().unwrap_or_default();
         let context_name = if let Some(name) = context {
             name
-        } else if let Some(name) = &config.current_context {
-            name
         } else {
-            return Err(KubeconfigError::CurrentContextNotSet);
+            current_context_name
         };
```
Comment (kazk) [inline kube-client/src/config/file_loader.rs:67]: Why was this changed?

File: kube-client/src/config/file_loader.rs [hunk 10]
```diff
@@ -59,40 +59,42 @@ impl ConfigLoader {
...
+
         let current_context = config
             .contexts
             .iter()
-            .find(|named_context| &named_context.name == context_name)
-            .map(|named_context| &named_context.context)
+            .find(|named_context| &named_context.name.clone().unwrap_or_default() == context_name)
```
Comment (kazk) [inline kube-client/src/config/file_loader.rs:72]:
```suggestion
            .find(|named_context| named_context.name.as_ref() == Some(context_name))
```

File: kube-client/src/config/file_loader.rs [hunk 11]
```diff
@@ -59,40 +59,42 @@ impl ConfigLoader {
...
+            .and_then(|named_context| named_context.context.clone())
             .ok_or_else(|| KubeconfigError::LoadContext(context_name.clone()))?;
 
-        let cluster_name = cluster.unwrap_or(&current_context.cluster);
+        let current_cluster_name = &current_context.cluster.clone().unwrap_or_default();
+        let cluster_name = cluster.unwrap_or(current_cluster_name);
```
Comment (kazk) [inline kube-client/src/config/file_loader.rs:77]: Should be something similar to how we get the context name:

```rust
        let cluster_name = if let Some(name) = cluster {
            name
        } else if let Some(name) = &current_context.cluster {
            name
        } else {
            return Err(KubeconfigError::CurrentClusterNotSet);
        };
```

Same for `user_name` below.

File: kube-client/src/config/file_loader.rs (hunk 9 above)
Comment (kazk) [reply kube-client/src/config/file_loader.rs:67]: I think we can clean this up to:

```rust
        let context_name = context
            .or_else(|| config.current_context.as_ref())
            .ok_or(KubeconfigError::CurrentContextNotSet)?;
```

Comment (goenning) [reply kube-client/src/config/file_loader.rs:67]: That was removed because current-context can be null/empty, at least that's how it works on kubectl. Do you think we should not change this in kube-rs?

Comment (kazk) [reply kube-client/src/config/file_loader.rs:67]: It can be null/empty when parsing, but should return an error when using the config, right?

Comment (goenning) [reply kube-client/src/config/file_loader.rs:67]: It depends, if we want similar behavior as client-go/kubectl, then an empty current-context is a valid value, because context names can be empty too.

File: kube-client/src/config/file_loader.rs (hunk 10 above)
Comment (goenning) [reply kube-client/src/config/file_loader.rs:72]: It's related to the comment above, if I apply this suggested change, then we it won't work when context name is null/none

File: kube-client/src/config/file_loader.rs (hunk 11 above)
Comment (goenning) [reply kube-client/src/config/file_loader.rs:77]: Same as the above, clusters can have empty names

File: kube-client/src/config/file_loader.rs (hunk 9 above)
Comment (goenning) [reply kube-client/src/config/file_loader.rs:67]: cluster and users can have empty names too

Comment (kazk) [reply kube-client/src/config/file_loader.rs:67]: client-go has [`ConfirmUsable`](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/client-go/tools/clientcmd/validation.go#L182) validation that's called [when creating a config for REST client](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L133) from kubeconfig. `ConfirmUsable` seems to have validation similar to what we have here and more.

"http://localhost:8080" is from [`ClusterDefaults`](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L44) which has a comment `DEPRECATED`.

Comment (goenning) [reply kube-client/src/config/file_loader.rs:67]: But do you have a context with an empty name? This for example, is valid and works on kubectl. I haven't really navigated client-go code to understand what is it doing, I'm testing it from a user perspective

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: ...
    server: https://kubernetes.docker.internal:6443
  name:
contexts:
- context:
    cluster: 
    user: 
  name: 
current-context:
kind: Config
preferences: {}
users:
- name:
  user:
    client-certificate-data: ...
    client-key-data: ...
```

Comment (kazk) [reply kube-client/src/config/file_loader.rs:67]: Confirmed kubectl works with a context with empty name when current context is empty.

We should align with client-go (library), and not kubectl (app), though.

---

- client-go's [`Config`](https://github.com/kubernetes/client-go/blob/1ac8d459351e21458fd1041f41e43403eadcbdba/tools/clientcmd/api/v1/types.go#L28) struct. (The link in `file_config.rs` is pointing to the struct after the conversion.)
- [conversion](https://github.com/kubernetes/client-go/blob/7697067af71046b18e03dbda04e01a5bb17f9809/tools/clientcmd/api/v1/conversion.go#L28) from array of named objects to map. Errors on duplicate name. Missing name defaults to empty string because it's Go (nullable uses a pointer type).
- client-go has a validation to [reject empty string context name](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/client-go/tools/clientcmd/validation.go#L196). kubectl seems to work.
- What to do when there's no match. client-go has validation to reject? kubectl seems to use a deprecated default?

We should probably use default value when missing instead of `Option` to align with client-go's [`Config`](https://github.com/kubernetes/client-go/blob/1ac8d459351e21458fd1041f41e43403eadcbdba/tools/clientcmd/api/v1/types.go#L28) as much as possible. Existing `Option<String>` fields should be fixed if it doesn't match client-go for consistency (not in this PR). We should also follow `omitempty` tags and skip serializing empty.

The problem is allowing `null`. Do we really need to? It just feels weird to me, and I don't think this happens in practice.

Comment (goenning) [reply kube-client/src/config/file_loader.rs:67]: Maybe kubectl doesn't call ConfirmUsable?

---

I think this topic requires a bit more discussion and I don't want to rush into a solution, but I'd like to get the other changes on this PR merged soon.

As empty/null names are very edge cases scenarios, I'd like to propose that:
- revert changes to "name" fields and keep them as required Strings
- create an issue to discuss how do we want to handle those scenarios and implement that on a subsequent PR

What you think?

Comment (kazk) [reply kube-client/src/config/file_loader.rs:67]: Yeah, let's do that.

Comment (clux) [reply kube-client/src/config/file_loader.rs:67]: It's possible that all this lax parsing is due to supporting config merges. Configs that are invalid in isolation (e.g. no current context) are fine in the merged setting (where another file sets the current-context).

File: kube-client/src/config/mod.rs (hunk 8 above)
Comment (goenning) [reply kube-client/src/config/mod.rs]: Done

File: kube-client/src/config/file_config.rs (hunk 5 above)
Comment (goenning) [reply kube-client/src/config/file_config.rs]: Done, formatting is fixed

File: kube-client/src/config/file_config.rs (hunk 7 above)
Comment (goenning) [reply kube-client/src/config/file_config.rs]: Reverted as discussed

File: kube-client/src/config/file_loader.rs (hunk 10 above)
Comment (goenning) [reply kube-client/src/config/file_loader.rs]: Resolving this, no longer relevant after reverting.

File: kube-client/src/config/file_loader.rs (hunk 11 above)
Comment (goenning) [reply kube-client/src/config/file_loader.rs]: Resolving this, no longer relevant after reverting.

File: kube-client/src/config/file_loader.rs (hunk 9 above)
Comment (goenning) [reply kube-client/src/config/file_loader.rs]: OK, I've reverted a bunch of changes so the PR is a lot simpler now. Still open for feedback! Thanks

File: kube-client/src/config/file_loader.rs [hunk 12]
```diff
@@ -66,33 +66,34 @@ impl ConfigLoader {
         } else {
             return Err(KubeconfigError::CurrentContextNotSet);
         };
+
```
Comment (kazk) [inline kube-client/src/config/file_loader.rs:69]: Let's revert this file as well because these are not affected by the changes.

File: kube-client/src/config/file_config.rs [hunk 13]
```diff
@@ -66,27 +66,30 @@ pub struct Preferences {
 #[cfg_attr(test, derive(PartialEq, Eq))]
 pub struct NamedExtension {
     /// Name of extension
-    pub name: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
```
Comment (kazk) [inline kube-client/src/config/file_config.rs:70]: This `name` should be reverted as well.

**Review by kazk — CHANGES_REQUESTED** [review]

File: kube-client/src/config/file_loader.rs (hunk 12 above)
Comment (goenning) [reply kube-client/src/config/file_loader.rs:69]: it actually is affected, because context, cluster and user are now `Option<T>`

**Review by kazk — APPROVED** [review]
